metacat-main/src/main/java/com/netflix/metacat/main/manager/ConnectorManager.java

```diff
@@ -103,31 +116,91 @@ public void addPlugin(final ConnectorPlugin connectorPlugin) {
     /**
      * Creates a connection for the given catalog.
      *
-     * @param catalogName      catalog name
-     * @param connectorType    connector type
      * @param connectorContext metacat connector properties
      */
-    public synchronized void createConnection(final String catalogName,
-                                              final String connectorType,
-                                              final ConnectorContext connectorContext) {
+    public synchronized void createConnection(final ConnectorContext connectorContext) {
         Preconditions.checkState(!stopped.get(), "ConnectorManager is stopped");
+        final String connectorType = connectorContext.getConnectorType();
+        final String catalogName = connectorContext.getCatalogName();
         final ConnectorPlugin connectorPlugin = plugins.get(connectorType);
         if (connectorPlugin != null) {
-            Preconditions
-                .checkState(!connectorFactories.containsKey(catalogName), "A connector %s already exists", catalogName);
-            final ConnectorFactory connectorFactory =
-                connectorPlugin.create(catalogName, connectorContext);
-            connectorFactories.put(catalogName, connectorFactory);
-
             final MetacatCatalogConfig catalogConfig =
-                MetacatCatalogConfig.createFromMapAndRemoveProperties(connectorType,
+                MetacatCatalogConfig.createFromMapAndRemoveProperties(connectorType, catalogName,
                     connectorContext.getConfiguration());
-            catalogs.put(catalogName, catalogConfig);
+            final List<String> databaseNames = catalogConfig.getSchemaWhitelist();
+            if (databaseNames.isEmpty()) {
+                Preconditions.checkState(!catalogs.contains(catalogName, EMPTY_STRING),
+                    "A catalog with name %s already exists", catalogName);
+            } else {
+                databaseNames.forEach(databaseName -> {
+                    Preconditions.checkState(!catalogs.contains(catalogName, databaseName),
+                        "A catalog with name %s for database %s already exists", catalogName, databaseName);
+                });
+            }
+            catalogConfigs.add(catalogConfig);
+            final ConnectorFactory connectorFactory = connectorPlugin.create(connectorContext);
+            try {
+                databaseServices.add(connectorFactory.getDatabaseService());
+            } catch (UnsupportedOperationException e) {
+                log.debug("Catalog {} doesn't support getDatabaseService. Ignoring.", catalogName);
```

Comment:
These error messages don't reflect possibility of a shard name?

Reply:
Yeah, we could pass in the shard name but since we are ignoring it, I did not add it.

Reply:
Will add.